### PR TITLE
[SPARK-47761][SQL] Oracle: Support reading AnsiIntervalTypes

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

In this PR, I proposed to add support for reading well-defined ANSI interval types from Oracle databases.

- NTERVAL YEAR [(year_precision)] TO MONTH
  - Stores a period of time in years and months, where year_precision is the number of digits in the YEAR datetime field. Accepted values are 0 to 9. The default is 2. The size is fixed at 5 bytes.
- INTERVAL DAY [(day_precision)] TO SECOND [(fractional_seconds_precision)]
  - Stores a period of time in days, hours, minutes, and seconds, where
    - day_precision is the maximum number of digits in the DAY datetime field. Accepted values are 0 to 9. The default is 2.
    - fractional_seconds_precision is the number of digits in the fractional part of the SECOND field. Accepted values are 0 to 9. The default is 6. The size is fixed at 11 bytes.

Both of them are mapped to the defaults of AnsiIntervalTypes.

We also add two developer APIs that convert interval strings to underlying representations of YearMonthIntervalType and DaytimeIntervalType. The default implementations assume that the inputs are compatible with the ANSI style, incompatible values will fail to be parsed. However, it shall fail much earlier at the data type mapping step because of undefined mapping rules

### Why are the changes needed?

Improve the Oracle accessibility, as tt's safe to read Oracle external intervals 

### Does this PR introduce _any_ user-facing change?

Yes, reading Oracle intervals are available now
### How was this patch tested?
new tests

### Was this patch authored or co-authored using generative AI tooling?
no

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @dongjoon-hyun, thank you .

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 1]
```diff
@@ -526,6 +526,16 @@ object JdbcUtils extends Logging with SQLConfHelper {
       (rs: ResultSet, row: InternalRow, pos: Int) =>
         row.update(pos, rs.getBytes(pos + 1))
 
+    case _: YearMonthIntervalType =>
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:529]: Is there any side-effect for the other JDBC dialects?

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala [hunk 2]
```diff
@@ -164,6 +166,17 @@ abstract class JdbcDialect extends Serializable with Logging {
...
+  def getYearMonthIntervalAsMonths(yearmonthStr: String): Int = {
+    fromYearMonthString(yearmonthStr).months
+  }
+
+  @Since("4.0.0")
+  def getDayTimeIntervalAsMicros(daytimeStr: String): Long = {
+    getDuration(fromDayTimeString(daytimeStr), TimeUnit.MICROSECONDS)
+  }
+
+
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:179]: nit. extra empty line.

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala [hunk 3]
```diff
@@ -540,4 +541,30 @@ class OracleIntegrationSuite extends DockerJDBCIntegrationSuite with SharedSpark
...
+      .option("query", query)
+      .load()
+    checkAnswer(df("SELECT INTERVAL '1-2' YEAR(1) TO MONTH as i0 FROM dual"),
+      Row(Period.of(1, 2, 0)))
+    checkAnswer(df("SELECT INTERVAL '1-2' YEAR(2) TO MONTH as i1 FROM dual"),
+      Row(Period.of(1, 2, 0)))
+    checkAnswer(df("SELECT INTERVAL '12345-2' YEAR(9) TO MONTH as i2 FROM dual"),
+      Row(Period.of(12345, 2, 0)))
+    checkAnswer(df("SELECT INTERVAL '1 12:23:56' DAY(1) TO SECOND(0) as i3 FROM dual"),
+        Row(Duration.ofDays(1).plusHours(12).plusMinutes(23).plusSeconds(56)))
```
Comment (dongjoon-hyun) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala:557]: nit. indentation?

**Comment by dongjoon-hyun** [comment]
Also, cc @gengliangwang , too.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 1 above)
Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:529]: It's safe here as there are no mapping rules to `YearMonthIntervalType` at the schema side

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (from my side)

**Comment by yaooqinn** [comment]
Thank you @dongjoon-hyun, I will make the CI all green and wait a while to merge if there's no objection from other reviewers

**Comment by yaooqinn** [comment]
Merged to master. Thank you very much @dongjoon-hyun

**Pull request closed without merging**